sdk/include/opentelemetry/sdk/trace/processor.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/trace/span.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+/**
+ * Span processor allow hooks for span start and end method invocations.
+ *
+ * Built-in span processors are responsible for batching and conversion of
+ * spans to exportable representation and passing batches to exporters.
+ */
+class SpanProcessor
+{
+public:
+  virtual ~SpanProcessor() = default;
+
+  // OnStart is called when a span is started.
+  virtual void OnStart(opentelemetry::trace::Span &span) noexcept = 0;
+
+  // OnEnd is called when a span is ended.
+  virtual void OnEnd(
+      nostd::shared_ptr<opentelemetry::sdk::trace::Recordable> &recordable) noexcept = 0;
+
+  // Export all ended spans that have not yet been exported.
+  virtual void ForceFlush(uint64_t timeout_ms) noexcept = 0;
+
+  // Shut down the processor and do any cleanup required.
```

Comment:
Does cleanup include flushing?

Reply:
Good question. In my reading the spec isn't clear on that, probably it's on us to decide? Java adds this to the documentation of this function:

> Implementations must ensure that all span events are processed before returning.

Reply:
Discussed in the Mar. 23 community meeting. Please update the comment (refer to the meeting notes) to clarify the semantic. Thanks!

Reply:
Done.